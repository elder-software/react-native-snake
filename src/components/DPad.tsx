import React from 'react';
import { View, TouchableHighlight } from 'react-native';
import FontAwesome from 'react-native-vector-icons/dist/FontAwesome';

interface DPadProps {
  onPressLeft: Function,
  onPressRight: Function,
  onPressUp: Function,
  onPressDown: Function,
}

const DPad: React.FC<DPadProps> = ({
  onPressLeft,
  onPressRight,
  onPressUp,
  onPressDown
}) => {
  return (
    <View style={{
      flex: 1,
      flexDirection: 'row',
      backgroundColor: 'white'
    }}>
      {/* Left arrow */}
      <TouchableHighlight
        onPress={() => onPressLeft()}
        style={{
          flex: 1,
          justifyContent: 'center'
        }}>
        <FontAwesome
          name="arrow-left"
          size={50}
          color="#900"
          style={{ alignSelf: 'flex-end' }}
        />
      </TouchableHighlight>


      {/* Middle up/down arrows */}
      <View style={{
        flex: 1,
        flexDirection: 'column'
      }}>
        {/* Up arrow */}
        <TouchableHighlight
          onPress={() => onPressUp()}
          style={{
            flex: 1,
            justifyContent: 'center'
          }}>
          <FontAwesome
            name="arrow-up"
            size={50}
            color="#900"
            style={{ alignSelf: 'center' }}
          />
        </TouchableHighlight>

        {/* Down arrow */}
        <TouchableHighlight
          onPress={() => onPressDown()}
          style={{
            flex: 1,
            justifyContent: 'center'
          }}>
          <FontAwesome
            name="arrow-down"
            size={50}
            color="#900"
            style={{ alignSelf: 'center' }}
          />
        </TouchableHighlight>
      </View>


      {/* Right arrow */}
      <TouchableHighlight
        onPress={() => onPressRight()}
        style={{
          flex: 1,
          justifyContent: 'center'
        }}>
        <FontAwesome
          name="arrow-right"
          size={50}
          color="#900"
          style={{ alignSelf: 'flex-start' }}
        />
      </TouchableHighlight>
    </View>
  );
}

export default DPad;