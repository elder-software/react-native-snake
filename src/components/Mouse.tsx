import React, { useState } from 'react';
import { View, Text, Dimensions } from 'react-native';


interface MouseProps {
  body: { position: { x: number, y: number } },
  color: any,
  size: number
}

const Mouse: React.FC<MouseProps> = ({ body, color, size }) => {
  const { position } = body;

  const x = position.x - size / 2;
  const y = position.y - size / 2;

  return (
    <View style={{
      position: 'absolute',
      width: size,
      height: size,
      left: x,
      top: y,
      backgroundColor: color,
      borderRadius: size
    }} />
  );
}


export default Mouse;
