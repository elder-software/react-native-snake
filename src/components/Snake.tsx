import React from 'react';
import { View } from 'react-native';

/**
 * Used to draw circles within the react native game engine and matter.js
 * Created using a simple View component
 * @param props used to set component props from the calling class
 */

interface SnakeProps {
  body: { position: { x: number, y: number } },
  color: any,
  size: number
}

const Snake: React.FC<SnakeProps> = ({ body, color, size }) => {
  const { position } = body;

  const x = position.x - size / 2;
  const y = position.y - size / 2;

  return (
    <View style={{
      position: 'absolute',
      width: size,
      height: size,
      left: x,
      top: y,
      backgroundColor: color,
      borderRadius: size
    }} />
  );
};



export default Snake;
