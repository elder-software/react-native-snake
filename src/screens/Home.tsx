import React, { useState, useEffect } from 'react';
import { View, Text, Dimensions } from 'react-native';
import { GameEngine } from "react-native-game-engine";
import Matter from "matter-js";
import DPad from '../components/DPad';
import { BALL_SIZE } from './constants';
import Snake from '../components/Snake';
import { BallMove } from './systems';

const screenDims = Dimensions.get('window');

const snakeSettings = {
  inertia: 0,
  friction: 0,
  frictionStatic: 0,
  frictionAir: 0,
  restitution: 1
};

export const snake = Matter.Bodies.circle(
  screenDims.width / 2,
  screenDims.width / 2,
  BALL_SIZE / 2,
  {
    ...snakeSettings,
    label: 'ball'
  }
);

export const mouse = Matter.Bodies.circle(
  screenDims.width / 2,
  screenDims.width / 2 + 30,
  BALL_SIZE / 2,
  {
    ...snakeSettings,
    label: 'mouse'
  }
);

// Creates the engine and world using Matter.js
const engine = Matter.Engine.create({ enableSleeping: false });
const world = engine.world;


// Adds all the objects to the world
Matter.World.add(world, [
  snake,
  mouse
]);

const Home: React.FC = () => {
  console.log(screenDims);

  useEffect(() => {
    Matter.Body.setVelocity(snake, { x: 1, y: 0 });

    Matter.Events.on(engine, 'collisionStart', event => {
      const pairs = event.pairs;

      const objA = pairs[0].bodyA.label;
      const objB = pairs[0].bodyB.label;

      if (objA == 'ball' && objB == 'mouse') {
        // Detects collision between the top wall and ball, player 1 scores
        // Increments player 1 score and restarts the ball with start position/velocity
        Matter.Body.setPosition(mouse, {
          x: 30,
          y: 30
        });
      }
    });
  });

  return (
    <View style={{ flex: 1 }}>
      <View style={{ width: screenDims.width, height: screenDims.width, backgroundColor: 'lightgrey' }}>
        <GameEngine
          systems={[BallMove]}
          entities={{
            physics: {
              engine: engine,
              world: world
            },
            snake: {
              body: snake,
              size: BALL_SIZE,
              color: '#ffffff',
              renderer: Snake
            },
            mouse: {
              body: mouse,
              size: BALL_SIZE,
              color: '#ffffff',
              renderer: Snake
            },
          }}>

        </GameEngine>
      </View>

      <View style={{ flex: 1 }}>
        <DPad
          onPressLeft={() => Matter.Body.setVelocity(snake, { x: -1, y: 0 })}
          onPressRight={() => Matter.Body.setVelocity(snake, { x: 1, y: 0 })}
          onPressUp={() => Matter.Body.setVelocity(snake, { x: 0, y: -1 })}
          onPressDown={() => Matter.Body.setVelocity(snake, { x: 0, y: 1 })}
        />
      </View>
    </View>
  );
}

export default Home;
