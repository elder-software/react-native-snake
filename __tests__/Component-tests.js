// import 'react-native';
import React from 'react';
import Home from '../src/screens/Home';
import DPad from '../src/components/DPad';
import Mouse from '../src/components/Mouse';
import Snake from '../src/components/Snake';

import renderer from 'react-test-renderer';

// test('Home Snapshot', () => {
//   const tree = renderer.create(<Home />).toJSON();
//   expect(tree).toMatchSnapshot();
// });

test('DPad Snapshot', () => {
  const tree = renderer.create(<DPad />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Mouse Snapshot', () => {
  const position = [100, 100];

  const tree = renderer.create(
    <Mouse
      body={{ position }}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Snake Snapshot', () => {
  const position = [100, 100];

  const tree = renderer.create(
    <Snake
      body={{ position }}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});