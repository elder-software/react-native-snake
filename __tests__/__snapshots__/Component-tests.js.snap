// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DPad Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#900",
            "fontSize": 50,
          },
          Object {
            "alignSelf": "flex-end",
          },
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#900",
              "fontSize": 50,
            },
            Object {
              "alignSelf": "center",
            },
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#900",
              "fontSize": 50,
            },
            Object {
              "alignSelf": "center",
            },
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#900",
            "fontSize": 50,
          },
          Object {
            "alignSelf": "flex-start",
          },
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Mouse Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "height": undefined,
      "left": NaN,
      "position": "absolute",
      "top": NaN,
      "width": undefined,
    }
  }
/>
`;

exports[`Snake Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "height": undefined,
      "left": NaN,
      "position": "absolute",
      "top": NaN,
      "width": undefined,
    }
  }
/>
`;
